samples/nrf9160/modem_shell/src/link/link_shell.c

```diff
@@ -1682,6 +1682,387 @@ int link_shell(const struct shell *shell, size_t argc, char **argv)
+					pdn_cid);
+				return -EINVAL;
+			}
+			break;
+		case '?':
+			goto show_usage;
+		default:
+			mosh_error("Unknown option. See usage:");
+			goto show_usage;
+		}
+	}
+
+	/* Check that all mandatory args were given: */
+	if (pdn_cid == 0) {
+		mosh_error("-I / --cid MUST be given. See usage:");
+		goto show_usage;
+	}
+
+	ret = link_shell_pdn_disconnect(pdn_cid);
+
+	return ret;
+
+show_usage:
+	/* Reset getopt for another users: */
+	optreset = 1;
+
+	link_shell_print_usage(LINK_CMD_DISCONNECT);
+	return ret;
+}
+
+int link_status(const struct shell *shell, size_t argc, char **argv)
+{
+	int ret = 0;
+	enum lte_lc_nw_reg_status current_reg_status;
+	enum lte_lc_func_mode functional_mode;
+	bool connected = false;
+	char snum[64];
+
+	ret = lte_lc_func_mode_get(&functional_mode);
+	if (ret) {
+		mosh_warn("Cannot get functional mode from modem: %d", ret);
+	} else {
+		mosh_print(
+			"Modem functional mode: %s",
+			link_shell_funmode_to_string(functional_mode, snum));
+	}
+	ret = lte_lc_nw_reg_status_get(&current_reg_status);
+	if (ret >= 0) {
+		link_shell_print_reg_status(current_reg_status);
+	} else {
+		mosh_error("Cannot get current registration status (%d)", ret);
+	}
+	if (current_reg_status == LTE_LC_NW_REG_REGISTERED_EMERGENCY ||
+		current_reg_status == LTE_LC_NW_REG_REGISTERED_HOME ||
+		current_reg_status == LTE_LC_NW_REG_REGISTERED_ROAMING) {
+		connected = true;
+	}
+
+	link_api_modem_info_get_for_shell(connected);
+	return ret;
+}
+
+enum link_shell_common_options link_shell_getopt_common(int argc, char *const argv[])
+{
+	enum link_shell_common_options common_option = LINK_COMMON_NONE;
+
+	/* Reset getopt due to possible previous failures: */
+	optreset = 1;
+
+	/* We start from subcmd arguments */
+	optind = 1;
+
+	int long_index = 0;
+	int opt;
+
+	while ((opt = getopt_long(argc, argv,
+				  "sured",
+				  long_options, &long_index)) != -1) {
+		switch (opt) {
+
+		case 's':
+			common_option = LINK_COMMON_SUBSCRIBE;
+			break;
+		case 'u':
+			common_option = LINK_COMMON_UNSUBSCRIBE;
+			break;
+		case 'e':
+			common_option = LINK_COMMON_ENABLE;
+			break;
+		case 'd':
+			common_option = LINK_COMMON_DISABLE;
+			break;
+		case 'r':
+			common_option = LINK_COMMON_READ;
+			break;
+		case LINK_SHELL_OPT_RESET:
+			common_option = LINK_COMMON_RESET;
+			break;
+		case LINK_SHELL_OPT_START:
+			common_option = LINK_COMMON_START;
+			break;
+		case LINK_SHELL_OPT_STOP:
+			common_option = LINK_COMMON_STOP;
+			break;
+		case '?':
+			/* TODO: We could print usage here */
+			break;
+		default:
+			break;
+		}
+	}
+	return common_option;
+}
+
+int link_funmode(const struct shell *shell, size_t argc, char **argv)
+{
+	int ret = 0;
+	bool nmode_use_rel14 = true;
+	enum lte_lc_func_mode funmode_option = LINK_FUNMODE_NONE;
+	enum lte_lc_func_mode functional_mode;
+	enum link_shell_common_options common_option;
+	char snum[64];
+
+	common_option = link_shell_getopt_common(argc, argv);
```

Comment:
This is one option of handling common options.
Drawback is that we call getopt_long twice for one command. Shouldn't matter but good to be aware of.

Reply:
Calling twice getopt_long is not an option as it permutates argv table. Implementation changed.